### ISSUE TITLE
Cannot allocate memory when creating indexes

### ISSUE THREAD

**Issue by neylsongularte** [meilisearch/meilisearch#1841]
I'm trying to create 10000+ indexes with 500+ documents and I get this error on a machine with 16GB of ram:

```
MeiliSearch ApiException: Http Status: 500 - Message: Cannot allocate memory (os error 12) - Error code: internal - Error type: internal_error - Error link: https://docs.meilisearch.com/errors#internal
```

![image](https://user-images.githubusercontent.com/4109823/138458881-08b247de-6587-4428-9a48-fccc64ee5106.png)

MeiliSearch version: 0.23.1

**Additional context**
Amazon Linux 2 - x86_64 - 16GB ram - With 10GB of ram free

1,1G	data.ms/

**Comment by tanmaymaheshwari97** [comment]
@neylsongularte Was this memory allocation issue fixed?

**Comment by curquiza** [comment]
Hello @tanmaymaheshwari97!
Can you check if you still have enough space on your machine for Meilisearch?

**Comment by tanmaymaheshwari97** [comment]
Hi @curquiza. My prod instance is on hosted on google cloud platform. And only around 4.5 GB of memory is utilized out of 16 GB.
![508A62CC-AA96-4B9A-BD31-C9097680A3AB_1_201_a](https://user-images.githubusercontent.com/66941891/175527243-e9c2b1aa-614d-4457-80f3-5a3835fc0a41.jpeg)

**Comment by curquiza** [comment]
Thanks for your quick answer. I was not asking about the RAM (16Gb), but the total about of space you have on your disk, and if there is enough space for Meilisearch to index your data 😊

**Comment by tanmaymaheshwari97** [comment]
I had a look at the disk space utilisation also. I don't think space is the issue here. Please suggest how to fix this issue as our search entirely depends on it.
![02C4D705-0E71-446C-AE14-FA5BC129B46E_1_201_a](https://user-images.githubusercontent.com/66941891/175536339-8d324adb-3d1a-4b90-a537-cd8bfb34df3c.jpeg)

**Comment by neylsongularte** [comment]
> I had a look at the disk space utilisation also. I don't think space is the issue here. Please suggest how to fix this issue as our search entirely depends on it. Thanks.

I changed the strategy and put it in a single index as suggested.

**Comment by curquiza** [comment]
For information, related issues: https://github.com/meilisearch/meilisearch/issues/1841
And it does not seem to be a huge amount of indexes

**Comment by dmitriid** [comment]
@curquiza I think you mean issue #2616 👍

**Comment by curquiza** [comment]
Discussed with @Kerollmops: an idea could be to use an LRU (least recently used) map instead of a HashMap

**Comment by dureuill** [comment]
> Discussed with @Kerollmops: an idea could be to use an LRU (least recently used) map instead of a HashMap

The issue with switching to a LRU is that the ecosystem does not have one that is maintained and can do `get()` operations behind a `RwLock`. I can roll one a simple one myself if we need to, though.

**Comment by curquiza** [comment]
Should be fixed by #3331, we will communicate here once the RC is out so that you can test it 😊
